### ISSUE TITLE
[STORM-183] Supervisor/worker shutdown hook should be called in distributed mode.

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#4025]
<p>if the process is killed forcefully from the OS or if it's crashing due to resource issues (e.g., out of memory), shutdown hooks won't be invoked.<br/>
-TERM (15) <br/>
The process is requested to stop running; it should try to exit cleanly <br/>
-KILL (9) <br/>
The process will be killed by the kernel; this signal cannot be ignored.<br/>
So should we better use 'kill -15' ?<br/>
See:<br/>
<a href="https://github.com/apache/incubator-storm/blob/master/storm-core/src/clj/backtype/storm/util.clj#L392" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/blob/master/storm-core/src/clj/backtype/storm/util.clj#L392</a></p>

<p><a href="https://github.com/apache/incubator-storm/blob/master/storm-core/src/clj/backtype/storm/daemon/supervisor.clj#L175" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/blob/master/storm-core/src/clj/backtype/storm/daemon/supervisor.clj#L175</a></p>

<p>will never be called for supervisor:<br/>
<a href="https://github.com/apache/incubator-storm/blob/master/storm-core/src/clj/backtype/storm/daemon/supervisor.clj#L396" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/blob/master/storm-core/src/clj/backtype/storm/daemon/supervisor.clj#L396</a></p>

<p>will never be called for worker:<br/>
<a href="https://github.com/apache/incubator-storm/blob/master/storm-core/src/clj/backtype/storm/daemon/worker.clj#L421" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/blob/master/storm-core/src/clj/backtype/storm/daemon/worker.clj#L421</a></p>

<p>We'd better add something like :<br/>
(.addShutdownHook (Runtime/getRuntime) (Thread. (fn [] (.shutdown mk-sv))))))  ?</p>

---
<details><summary><i>Originally reported by <a title="caofangkun" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=caofangkun">caofangkun</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-183" target="_blank">Supervisor/worker shutdown hook should be called in distributed mode.</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Parth Brahmbhatt" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=parth.brahmbhatt">parth.brahmbhatt</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2014-06-25T15:04:11+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=caofangkun">caofangkun</a>:</i>
<p>After the patch:<br/>
kill -15 supervisor_pid <br/>
tail -20f $STORM_HOME/logs/supervisor.log <br/>
2013-12-20 10:34:47 b.s.d.supervisor <span class="error">&#91;INFO&#93;</span> Shutting down supervisor ae96c39c-5e62-4e88-824d-3e5b05ac31bf<br/>
2013-12-20 10:34:47 b.s.event <span class="error">&#91;INFO&#93;</span> Event manager interrupted<br/>
2013-12-20 10:34:47 b.s.event <span class="error">&#91;INFO&#93;</span> Event manager interrupted<br/>
2013-12-20 10:34:47 o.a.z.ZooKeeper <span class="error">&#91;INFO&#93;</span> Session: 0x1430dbb25c00027 closed<br/>
2013-12-20 10:34:47 o.a.z.ClientCnxn <span class="error">&#91;INFO&#93;</span> EventThread shut down</p>


<p>kill -9 worker_pid <br/>
tail -100f $STORM_HOME/logs/worker-6101.log <br/>
2013-12-20 10:35:01 b.s.d.worker <span class="error">&#91;INFO&#93;</span> Shutting down receive thread<br/>
2013-12-20 10:35:01 b.s.m.loader <span class="error">&#91;INFO&#93;</span> Shutting down receiving-thread: <span class="error">&#91;WordCountTopology-1-1387364272, 39931&#93;</span><br/>
2013-12-20 10:35:01 b.s.m.loader <span class="error">&#91;INFO&#93;</span> Receiving-thread:<span class="error">&#91;WordCountTopology-1-1387364272, 39931&#93;</span> received shutdown notice<br/>
2013-12-20 10:35:01 b.s.m.loader <span class="error">&#91;INFO&#93;</span> Waiting for receiving-thread:<span class="error">&#91;WordCountTopology-1-1387364272, 39931&#93;</span> to die<br/>
2013-12-20 10:35:01 b.s.m.loader <span class="error">&#91;INFO&#93;</span> Shutdown receiving-thread: <span class="error">&#91;WordCountTopology-1-1387364272, 39931&#93;</span><br/>
2013-12-20 10:35:01 b.s.d.worker <span class="error">&#91;INFO&#93;</span> Shut down receive thread<br/>
2013-12-20 10:35:01 b.s.d.worker <span class="error">&#91;INFO&#93;</span> Terminating messaging context<br/>
2013-12-20 10:35:01 b.s.d.worker <span class="error">&#91;INFO&#93;</span> Shutting down executors<br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shutting down executor __acker:<span class="error">&#91;3 3&#93;</span><br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shut down executor __acker:<span class="error">&#91;3 3&#93;</span><br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shutting down executor count:<span class="error">&#91;7 7&#93;</span><br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shut down executor count:<span class="error">&#91;7 7&#93;</span><br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shutting down executor count:<span class="error">&#91;11 11&#93;</span><br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shut down executor count:<span class="error">&#91;11 11&#93;</span><br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shutting down executor count:<span class="error">&#91;15 15&#93;</span><br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shut down executor count:<span class="error">&#91;15 15&#93;</span><br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shutting down executor split:<span class="error">&#91;19 19&#93;</span><br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shut down executor split:<span class="error">&#91;19 19&#93;</span><br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shutting down executor split:<span class="error">&#91;23 23&#93;</span><br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shut down executor split:<span class="error">&#91;23 23&#93;</span><br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shutting down executor spout:<span class="error">&#91;27 27&#93;</span><br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shut down executor spout:<span class="error">&#91;27 27&#93;</span><br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shutting down executor __system:<span class="error">&#91;-1 -1&#93;</span><br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.d.executor <span class="error">&#91;INFO&#93;</span> Shut down executor __system:<span class="error">&#91;-1 -1&#93;</span><br/>
2013-12-20 10:35:01 b.s.d.worker <span class="error">&#91;INFO&#93;</span> Shut down executors<br/>
2013-12-20 10:35:01 b.s.d.worker <span class="error">&#91;INFO&#93;</span> Shutting down transfer thread<br/>
2013-12-20 10:35:01 b.s.util <span class="error">&#91;INFO&#93;</span> Async loop interrupted!<br/>
2013-12-20 10:35:01 b.s.d.worker <span class="error">&#91;INFO&#93;</span> Shut down transfer thread<br/>
2013-12-20 10:35:01 b.s.d.worker <span class="error">&#91;INFO&#93;</span> Shutting down default resources<br/>
2013-12-20 10:35:01 b.s.d.worker <span class="error">&#91;INFO&#93;</span> Shut down default resources<br/>
2013-12-20 10:35:01 b.s.d.worker <span class="error">&#91;INFO&#93;</span> Disconnecting from storm cluster state context<br/>
2013-12-20 10:35:01 o.a.z.ClientCnxn <span class="error">&#91;INFO&#93;</span> EventThread shut down<br/>
2013-12-20 10:35:01 o.a.z.ZooKeeper <span class="error">&#91;INFO&#93;</span> Session: 0x1430dbb25c00031 closed<br/>
2013-12-20 10:35:01 b.s.d.worker <span class="error">&#91;INFO&#93;</span> Shut down worker WordCountTopology-1-1387364272 ae96c39c-5e62-4e88-824d-3e5b05ac31bf 39931</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p>My only issue here is that we really would want to back up the kill -15 with a kill -9 a little while later.  We cannot be sure that the process is down unless we use the -9.  But this would either mean that we have to wait a certain amount of time no matter what in between the two calls or we need a way to check if the process is still up.  I can see the check being added in later if needed, but I don't want to rely on user code to always exit cleanly when asked politely.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user Parth-Brahmbhatt opened a pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139</a></p>

<p>     <a href="https://issues.apache.org/jira/browse/STORM-183" title="Supervisor/worker shutdown hook should be called in distributed mode." class="issue-link" data-issue-key="STORM-183"><del>STORM-183</del></a>: supervisor and worker should rugister with runtime for shutd...</p>

<p>    ...own hooks. Supervisors should use kill -15 before kill -9 to allow workers to perform cleanup.</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/Parth-Brahmbhatt/incubator-storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/Parth-Brahmbhatt/incubator-storm</a> <a href="https://issues.apache.org/jira/browse/STORM-183" title="Supervisor/worker shutdown hook should be called in distributed mode." class="issue-link" data-issue-key="STORM-183"><del>STORM-183</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #139</p>

<hr />
<p>commit f15714c0a8342147c9cfe4410c86c0a325d30a4b<br/>
Author: Parth-Brahmbhatt <pbrahmbhatt@hw10960.local><br/>
Date:   2014-06-12T22:54:34Z</p>

<p>     <a href="https://issues.apache.org/jira/browse/STORM-183" title="Supervisor/worker shutdown hook should be called in distributed mode." class="issue-link" data-issue-key="STORM-183"><del>STORM-183</del></a>: supervisor and worker should rugister with runtime for shutdown hooks. Supervisors should use kill -15 before kill -9 to allow workers to perform cleanup.</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139#issuecomment-46036698" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139#issuecomment-46036698</a></p>

<p>    Some changes were not pushed with the last commit so pushed them and resending now.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139#discussion_r13763330" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139#discussion_r13763330</a></p>

<p>    &#8212; Diff: storm-core/src/clj/backtype/storm/util.clj &#8212;<br/>
    @@ -412,6 +404,41 @@<br/>
     (defn sleep-until-secs <span class="error">&#91;target-secs&#93;</span><br/>
       (Time/sleepUntil (* (long target-secs) 1000)))</p>

<p>    +(def ^:const sig-kill 9)<br/>
    +<br/>
    +(def ^:const sig-term 15)<br/>
    +<br/>
    +(defn send-signal-to-process <span class="error">&#91;pid signum&#93;</span><br/>
    &#8212; End diff &#8211;</p>

<p>    With the new coding guidelines, could you put the parameter list on a separate line?</p>

<p>    ```<br/>
    (defn send-signal-to-process<br/>
      <span class="error">&#91;pid signum&#93;</span><br/>
      ...)<br/>
    ```</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139#discussion_r13765591" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139#discussion_r13765591</a></p>

<p>    &#8212; Diff: storm-core/src/clj/backtype/storm/daemon/worker.clj &#8212;<br/>
    @@ -448,4 +448,6 @@<br/>
     (defn -main <span class="error">&#91;storm-id assignment-id port-str worker-id&#93;</span>  <br/>
       (let <span class="error">&#91;conf (read-storm-config)&#93;</span><br/>
 (validate-distributed-mode! conf)</p>
<ul class="alternate" type="square">
	<li>(mk-worker conf nil storm-id assignment-id (Integer/parseInt port-str) worker-id)))<br/>
    +    (.addShutdownHook (Runtime/getRuntime)
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    I'm not sure that this shutdown hook will be called in all cases.  Most places the worker code calls Runtime.halt on error, which does not call any of the shutdown hooks.  It just exits.  I think that was done because shutdown hooks can lock up, and if they do the process does not exit.  So we could get zombie processes.  Typically I have seen this addressed by installing a shutdown hook that will let the other shutdown hooks run for a maximum amount of time, at which point it calls Runtime.halt.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139#discussion_r13766558" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139#discussion_r13766558</a></p>

<p>    &#8212; Diff: storm-core/src/clj/backtype/storm/util.clj &#8212;<br/>
    @@ -412,6 +404,41 @@<br/>
     (defn sleep-until-secs <span class="error">&#91;target-secs&#93;</span><br/>
       (Time/sleepUntil (* (long target-secs) 1000)))</p>

<p>    +(def ^:const sig-kill 9)<br/>
    +<br/>
    +(def ^:const sig-term 15)<br/>
    +<br/>
    +(defn send-signal-to-process <span class="error">&#91;pid signum&#93;</span><br/>
    &#8212; End diff &#8211;</p>

<p>    adding this, can you also point me to the guidelines you are referring to.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139#discussion_r13768469" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139#discussion_r13768469</a></p>

<p>    &#8212; Diff: storm-core/src/clj/backtype/storm/util.clj &#8212;<br/>
    @@ -412,6 +404,41 @@<br/>
     (defn sleep-until-secs <span class="error">&#91;target-secs&#93;</span><br/>
       (Time/sleepUntil (* (long target-secs) 1000)))</p>

<p>    +(def ^:const sig-kill 9)<br/>
    +<br/>
    +(def ^:const sig-term 15)<br/>
    +<br/>
    +(defn send-signal-to-process <span class="error">&#91;pid signum&#93;</span><br/>
    &#8212; End diff &#8211;</p>

<p>    This was just recently merged in. The guidelines are under<br/>
    <a href="https://github.com/apache/incubator-storm/blob/master/DEVELOPER.md#contribute-code" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/blob/master/DEVELOPER.md#contribute-code</a><br/>
    In the paragraph starting with 'Please use idiomatic Clojure style'</p>

<p>    For the most part it is place parameters in a separate line from the function name.  Use document strings as needed.  Don't have closing ')' on their own line.  I know we don't follow it everywhere, but we are trying to start.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139#issuecomment-46065430" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139#issuecomment-46065430</a></p>

<p>    addressed comments , Would appreciate any suggestions for a better and less verbose name for function add-shutdown-hook-with-force-kill-in-1-sec.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139#issuecomment-46197891" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139#issuecomment-46197891</a></p>

<p>    There are a number of merge conflicts now.  Could you please upmerge to master.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139#issuecomment-46238986" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139#issuecomment-46238986</a></p>

<p>    Ran into some weird git issues so just forked off a new repo and created a new pull request, <a href="https://github.com/apache/incubator-storm/pull/143" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143</a>. The fork was created like an hour ago so should not have any merge conflicts. I can merge to my master if it makes any easier for you to review.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139#issuecomment-46241120" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139#issuecomment-46241120</a></p>

<p>    Could you close this pull request then, and add <a href="https://issues.apache.org/jira/browse/STORM-183" title="Supervisor/worker shutdown hook should be called in distributed mode." class="issue-link" data-issue-key="STORM-183"><del>STORM-183</del></a> to the other one?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt closed the pull request at:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139#issuecomment-46241965" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139#issuecomment-46241965</a></p>

<p>    closed this one, not sure what you mean by add <a href="https://issues.apache.org/jira/browse/STORM-183" title="Supervisor/worker shutdown hook should be called in distributed mode." class="issue-link" data-issue-key="STORM-183"><del>STORM-183</del></a>, do you mean adding "<a href="https://issues.apache.org/jira/browse/STORM-183" title="Supervisor/worker shutdown hook should be called in distributed mode." class="issue-link" data-issue-key="STORM-183"><del>STORM-183</del></a>" text as prefix to pull request and commit message?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#discussion_r13833165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#discussion_r13833165</a></p>

<p>    &#8212; Diff: storm-core/src/clj/backtype/storm/util.clj &#8212;<br/>
    @@ -388,13 +388,51 @@<br/>
 (catch IOException e<br/>
   (log-message "Could not extract " dir " from " jarpath))))</p>

<p>    -(defn ensure-process-killed! <span class="error">&#91;pid&#93;</span><br/>
    +(defn sleep-secs <span class="error">&#91;secs&#93;</span><br/>
    +  (when (pos? secs)<br/>
    +    (Time/sleep (* (long secs) 1000))))<br/>
    +<br/>
    +(defn sleep-until-secs <span class="error">&#91;target-secs&#93;</span><br/>
    +  (Time/sleepUntil (* (long target-secs) 1000)))<br/>
    +<br/>
    +(def ^:const sig-kill 9)<br/>
    +<br/>
    +(def ^:const sig-term 15)<br/>
    +<br/>
    +(defn send-signal-to-process<br/>
    +  <span class="error">&#91;pid signum&#93;</span><br/>
    +  (try-cause<br/>
    +    (exec-command! (str (if on-windows?<br/>
    +  (if (== signum sig-kill) "taskkill /f /pid " "taskkill /pid ")<br/>
    +  (str "kill -" signum " "))<br/>
    +     pid))<br/>
    +    (catch ExecuteException e<br/>
    +      (log-message "Error when trying to kill " pid ". Process is probably already dead."))))<br/>
    +<br/>
    +(defn force-kill-process<br/>
    +  <span class="error">&#91;pid&#93;</span><br/>
    +  (send-signal-to-process pid sig-kill))<br/>
    +<br/>
    +(defn kill-process-with-sig-term<br/>
    +  <span class="error">&#91;pid&#93;</span><br/>
    +  (send-signal-to-process pid sig-term))<br/>
    +<br/>
    +(defn ensure-process-killed!<br/>
    +  <span class="error">&#91;pid&#93;</span><br/>
       ;; TODO: should probably do a ps ax of some sort to make sure it was killed<br/>
       (try-cause</p>
<ul class="alternate" type="square">
	<li>(exec-command! (str (if on-windows? "taskkill /f /pid " "kill -9 ") pid))<br/>
    +    (kill-process-with-sig-term pid)<br/>
 (catch ExecuteException e<br/>
   (log-message "Error when trying to kill " pid ". Process is probably already dead."))))</li>
</ul>


<p>    +(defn add-shutdown-hook-with-force-kill-in-1-sec<br/>
    &#8212; End diff &#8211;</p>

<p>    I was thinking that this would not run some other function before calling sleep.  The Shutdown hook runs when the process is going to exit, if the other function gets stuck (which I was able to do in my testing) then the process never exits.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#discussion_r13833188" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#discussion_r13833188</a></p>

<p>    &#8212; Diff: storm-core/src/clj/backtype/storm/util.clj &#8212;<br/>
    @@ -388,13 +388,51 @@<br/>
 (catch IOException e<br/>
   (log-message "Could not extract " dir " from " jarpath))))</p>

<p>    -(defn ensure-process-killed! <span class="error">&#91;pid&#93;</span><br/>
    +(defn sleep-secs <span class="error">&#91;secs&#93;</span><br/>
    +  (when (pos? secs)<br/>
    +    (Time/sleep (* (long secs) 1000))))<br/>
    +<br/>
    +(defn sleep-until-secs <span class="error">&#91;target-secs&#93;</span><br/>
    +  (Time/sleepUntil (* (long target-secs) 1000)))<br/>
    +<br/>
    +(def ^:const sig-kill 9)<br/>
    +<br/>
    +(def ^:const sig-term 15)<br/>
    +<br/>
    +(defn send-signal-to-process<br/>
    +  <span class="error">&#91;pid signum&#93;</span><br/>
    +  (try-cause<br/>
    +    (exec-command! (str (if on-windows?<br/>
    +  (if (== signum sig-kill) "taskkill /f /pid " "taskkill /pid ")<br/>
    +  (str "kill -" signum " "))<br/>
    +     pid))<br/>
    +    (catch ExecuteException e<br/>
    +      (log-message "Error when trying to kill " pid ". Process is probably already dead."))))<br/>
    +<br/>
    +(defn force-kill-process<br/>
    +  <span class="error">&#91;pid&#93;</span><br/>
    +  (send-signal-to-process pid sig-kill))<br/>
    +<br/>
    +(defn kill-process-with-sig-term<br/>
    +  <span class="error">&#91;pid&#93;</span><br/>
    +  (send-signal-to-process pid sig-term))<br/>
    +<br/>
    +(defn ensure-process-killed!<br/>
    +  <span class="error">&#91;pid&#93;</span><br/>
       ;; TODO: should probably do a ps ax of some sort to make sure it was killed<br/>
       (try-cause</p>
<ul class="alternate" type="square">
	<li>(exec-command! (str (if on-windows? "taskkill /f /pid " "kill -9 ") pid))<br/>
    +    (kill-process-with-sig-term pid)<br/>
 (catch ExecuteException e<br/>
   (log-message "Error when trying to kill " pid ". Process is probably already dead."))))</li>
</ul>


<p>    +(defn add-shutdown-hook-with-force-kill-in-1-sec<br/>
    +  "adds the user supplied function as a shutdown hook for cleanup.<br/>
    +   Also adds a function that sleeps for a second and then sends kill -9 to process to avoid any zombie process in case<br/>
    +   cleanup function hangs."<br/>
    +  <span class="error">&#91;func&#93;</span><br/>
    +  (.addShutdownHook (Runtime/getRuntime) (Thread. #((func)<br/>
    +     (sleep-secs 1)<br/>
    +     (force-kill-process (process-pid))))))<br/>
    &#8212; End diff &#8211;</p>

<p>    I would prefer to do a Runtime.halt here instead of force-kill-process.  I think it is more future proof.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user ptgoetz commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/139#issuecomment-46244421" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/139#issuecomment-46244421</a></p>

<p>    Yes, just prefix the title with the JIRA ID.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#discussion_r13833622" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#discussion_r13833622</a></p>

<p>    &#8212; Diff: storm-core/src/clj/backtype/storm/util.clj &#8212;<br/>
    @@ -388,13 +388,51 @@<br/>
 (catch IOException e<br/>
   (log-message "Could not extract " dir " from " jarpath))))</p>

<p>    -(defn ensure-process-killed! <span class="error">&#91;pid&#93;</span><br/>
    +(defn sleep-secs <span class="error">&#91;secs&#93;</span><br/>
    +  (when (pos? secs)<br/>
    +    (Time/sleep (* (long secs) 1000))))<br/>
    +<br/>
    +(defn sleep-until-secs <span class="error">&#91;target-secs&#93;</span><br/>
    +  (Time/sleepUntil (* (long target-secs) 1000)))<br/>
    +<br/>
    +(def ^:const sig-kill 9)<br/>
    +<br/>
    +(def ^:const sig-term 15)<br/>
    +<br/>
    +(defn send-signal-to-process<br/>
    +  <span class="error">&#91;pid signum&#93;</span><br/>
    +  (try-cause<br/>
    +    (exec-command! (str (if on-windows?<br/>
    +  (if (== signum sig-kill) "taskkill /f /pid " "taskkill /pid ")<br/>
    +  (str "kill -" signum " "))<br/>
    +     pid))<br/>
    +    (catch ExecuteException e<br/>
    +      (log-message "Error when trying to kill " pid ". Process is probably already dead."))))<br/>
    +<br/>
    +(defn force-kill-process<br/>
    +  <span class="error">&#91;pid&#93;</span><br/>
    +  (send-signal-to-process pid sig-kill))<br/>
    +<br/>
    +(defn kill-process-with-sig-term<br/>
    +  <span class="error">&#91;pid&#93;</span><br/>
    +  (send-signal-to-process pid sig-term))<br/>
    +<br/>
    +(defn ensure-process-killed!<br/>
    +  <span class="error">&#91;pid&#93;</span><br/>
       ;; TODO: should probably do a ps ax of some sort to make sure it was killed<br/>
       (try-cause</p>
<ul class="alternate" type="square">
	<li>(exec-command! (str (if on-windows? "taskkill /f /pid " "kill -9 ") pid))<br/>
    +    (kill-process-with-sig-term pid)<br/>
 (catch ExecuteException e<br/>
   (log-message "Error when trying to kill " pid ". Process is probably already dead."))))</li>
</ul>


<p>    +(defn add-shutdown-hook-with-force-kill-in-1-sec<br/>
    &#8212; End diff &#8211;</p>

<p>    You are right, I want to avoid a model where people have to call (.addShutdownHook blah blah) and then remember to call (add-shutdown-hook-with-force-kill-in-1-sec). <br/>
    How about adding the use function and the actual forcing function as two seperate shutdown hook threads? <br/>
    (defn add-shutdown-hook-with-force-kill-in-1-sec<br/>
      <span class="error">&#91;func&#93;</span><br/>
      (.addShutdownHook (Runtime/getRuntime) (Thread. #(func)))<br/>
      (.addShutdownHook (Runtime/getRuntime) (Thread. #((sleep-secs 1)<br/>
(.halt (RunTime/getRunTime) 20)))))</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#discussion_r13833623" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#discussion_r13833623</a></p>

<p>    &#8212; Diff: storm-core/src/clj/backtype/storm/util.clj &#8212;<br/>
    @@ -388,13 +388,51 @@<br/>
 (catch IOException e<br/>
   (log-message "Could not extract " dir " from " jarpath))))</p>

<p>    -(defn ensure-process-killed! <span class="error">&#91;pid&#93;</span><br/>
    +(defn sleep-secs <span class="error">&#91;secs&#93;</span><br/>
    +  (when (pos? secs)<br/>
    +    (Time/sleep (* (long secs) 1000))))<br/>
    +<br/>
    +(defn sleep-until-secs <span class="error">&#91;target-secs&#93;</span><br/>
    +  (Time/sleepUntil (* (long target-secs) 1000)))<br/>
    +<br/>
    +(def ^:const sig-kill 9)<br/>
    +<br/>
    +(def ^:const sig-term 15)<br/>
    +<br/>
    +(defn send-signal-to-process<br/>
    +  <span class="error">&#91;pid signum&#93;</span><br/>
    +  (try-cause<br/>
    +    (exec-command! (str (if on-windows?<br/>
    +  (if (== signum sig-kill) "taskkill /f /pid " "taskkill /pid ")<br/>
    +  (str "kill -" signum " "))<br/>
    +     pid))<br/>
    +    (catch ExecuteException e<br/>
    +      (log-message "Error when trying to kill " pid ". Process is probably already dead."))))<br/>
    +<br/>
    +(defn force-kill-process<br/>
    +  <span class="error">&#91;pid&#93;</span><br/>
    +  (send-signal-to-process pid sig-kill))<br/>
    +<br/>
    +(defn kill-process-with-sig-term<br/>
    +  <span class="error">&#91;pid&#93;</span><br/>
    +  (send-signal-to-process pid sig-term))<br/>
    +<br/>
    +(defn ensure-process-killed!<br/>
    +  <span class="error">&#91;pid&#93;</span><br/>
       ;; TODO: should probably do a ps ax of some sort to make sure it was killed<br/>
       (try-cause</p>
<ul class="alternate" type="square">
	<li>(exec-command! (str (if on-windows? "taskkill /f /pid " "kill -9 ") pid))<br/>
    +    (kill-process-with-sig-term pid)<br/>
 (catch ExecuteException e<br/>
   (log-message "Error when trying to kill " pid ". Process is probably already dead."))))</li>
</ul>


<p>    +(defn add-shutdown-hook-with-force-kill-in-1-sec<br/>
    +  "adds the user supplied function as a shutdown hook for cleanup.<br/>
    +   Also adds a function that sleeps for a second and then sends kill -9 to process to avoid any zombie process in case<br/>
    +   cleanup function hangs."<br/>
    +  <span class="error">&#91;func&#93;</span><br/>
    +  (.addShutdownHook (Runtime/getRuntime) (Thread. #((func)<br/>
    +     (sleep-secs 1)<br/>
    +     (force-kill-process (process-pid))))))<br/>
    &#8212; End diff &#8211;</p>

<p>    done.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-46245056" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-46245056</a></p>

<p>    I have also hit a few situations where the kill -9 never came out.  I saw the kill -15 but not the final death blow.  I can reproduce it by launching several word count applications on a 1 node cluster, and then killing them 1 at a time (whichever one has the full 3 workers running)  Eventually I am left with several workers that refuse to die.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-46246787" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-46246787</a></p>

<p>    i am trying to reproduce this but in my case the worker always ends up killing, I am testing with the new change where the force-shutdown runs in a separate thread so it is not blocked by the other shutdown hook function. How r u killing the workers? Did you make any other code changes locally to ensure workers hang around longer?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-46248380" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-46248380</a></p>

<p>    When I ran I did run jstack on the workers and all of them were stuck somewhere in the shutdown.  So if you fixed the worker you would not see this.</p>

<p>    ```"Thread-23" #51 prio=5 os_prio=31 tid=0x00007f91a1059000 nid=0xb103 in Object.wait() <span class="error">&#91;0x0000000193600000&#93;</span><br/>
       java.lang.Thread.State: WAITING (on object monitor)<br/>
    	at java.lang.Object.wait(Native Method)<br/>
    	at java.lang.Thread.join(Thread.java:1244)</p>
<ul class="alternate" type="square">
	<li>locked <0x000000011a65d4f8> (a java.lang.Thread)<br/>
    	at java.lang.Thread.join(Thread.java:1318)<br/>
    	at backtype.storm.util$async_loop$reify__1051.join(util.clj:486)<br/>
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)<br/>
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)<br/>
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)<br/>
    	at java.lang.reflect.Method.invoke(Method.java:483)<br/>
    	at clojure.lang.Reflector.invokeMatchingMethod(Reflector.java:93)<br/>
    	at clojure.lang.Reflector.invokeNoArgInstanceMember(Reflector.java:298)<br/>
    	at backtype.storm.daemon.executor$mk_executor$reify__5504.shutdown(executor.clj:353)<br/>
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)<br/>
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)<br/>
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)<br/>
    	at java.lang.reflect.Method.invoke(Method.java:483)<br/>
    	at clojure.lang.Reflector.invokeMatchingMethod(Reflector.java:93)<br/>
    	at clojure.lang.Reflector.invokeNoArgInstanceMember(Reflector.java:298)<br/>
    	at backtype.storm.daemon.worker$fn_<em>5930$exec_fn</em><em>2806</em><em>auto</em>__<em>5931$shutdown_STAR</em>__5949.invoke(worker.clj:392)<br/>
    	at backtype.storm.daemon.worker$fn_<em>5930$exec_fn</em><em>2806</em><em>auto</em><em>$reify</em>_5975.shutdown(worker.clj:423)<br/>
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)<br/>
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)<br/>
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)<br/>
    	at java.lang.reflect.Method.invoke(Method.java:483)<br/>
    	at clojure.lang.Reflector.invokeMatchingMethod(Reflector.java:93)<br/>
    	at clojure.lang.Reflector.invokeNoArgInstanceMember(Reflector.java:298)<br/>
    	at backtype.storm.daemon.worker$<em>main$fn</em>_6002.invoke(worker.clj:455)<br/>
    	at backtype.storm.util$add_shutdown_hook_with_force_kill_in_1_sec$fn__975.invoke(util.clj:433)<br/>
    	at clojure.lang.AFn.run(AFn.java:24)<br/>
    	at java.lang.Thread.run(Thread.java:744)```</li>
</ul>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-46250710" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-46250710</a></p>

<p>    I have tried with both old and new code and I still can not reproduce this. I have tried killing the process with both SIGTERM and SIGINT, in both cases the process terminates.</p>

<p>    If you are killing in some other way or if you have some other local code change to make workers sleep during shutdown phase then let me know. </p>

<p>    I have checked in the change to make the final shutdown blow run in a separate thread.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-46302266" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-46302266</a></p>

<p>    I do like combining the two shutdown hooks together.</p>

<p>    I understand what is happening now with the leaked processes now.  I was confused about exactly what the code was supposed to be doing.</p>

<p>    I thought that the supervisor would send a sig term to the process, and then wait a while and send a sig kill to it.  I didn't realize that instead it was just sending a sigterm, and letting the worker send the sigkill to itself.</p>

<p>    The issue with not having the supervisor force-kill the child is that a bug in the worker, or in a child process the worker forks, could result in the process being leaked.  </p>

<p>    I don't want to do the sleep right after the sigterm, because if there are multiple workers the sleeps will add up.  I think we want to modify sync-processes in the supervisor to do 2 passes over the workers that need to be killed.  The first pass would ask the worker to exit (sigterm).  The second pass would force-kill the worker and cleanup the directories associated with it.  There could be a 1 second sleep in between if any workers were being killed (I don't want to sleep if no workers are shutting down). </p>

<p>    Does that sound reasonable?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-46333200" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-46333200</a></p>

<p>    Not sure why a bug in the worker would cause leaked processes even when we already have a shutdown hook in workers that should kill it in a second but I am no expert on unix  signal processing so I will trust you on that. Added the supervisor sequence to do what you have proposed. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p>It is mostly to cover all possible situations.  In unix a process can install a handler for SIGTERM and actually ignore it entirely.  But SIGKILL cannot be ignored.  The OS will not give the user process any chance to handle it.  Java handles SIGTERM and calls System.exit when it happens.  But if the shutdown handlers lock up for some reason we have effectively ignored the SIGTERM.  In my case something happened so the executor would not exit, not sure why, but this caused the shutdown to never finish. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-46592535" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-46592535</a></p>

<p>    @revans2  When you have a chance can you please review the latest change? I have implemented what you proposed in your last comment.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-46615057" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-46615057</a></p>

<p>    Thanks for the update.  It looks good, but a few minor things to rearrange. Thanks for sticking with this it will be a huge benefit to storm once it is in.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-46622640" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-46622640</a></p>

<p>    Thanks for clarifying that, I forgot that is what it was doing. Looking forward to your update reordering the supervisor cleanup order.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-46717551" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-46717551</a></p>

<p>    Superviosr cleanup order has been updated. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-46732024" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-46732024</a></p>

<p>    I am a +1.  Everything seems to be working OK.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user ptgoetz commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-47037071" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-47037071</a></p>

<p>    +1</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143#issuecomment-47103756" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143#issuecomment-47103756</a></p>

<p>    @Parth-Brahmbhatt I am going to merge this in, but I want to make sure I give the proper credit, both in the README and on the JIRA.  For now I am just including your github account, but if you want me to assign the JIRA to you and update the README please let me know which JIRA account to use.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/143" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/143</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p>I merged this into master, Thanks Parth-Brahmbhatt for the pull request.</p>

**Comment by jira-importer** [comment]
<i>[Originally related to: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-126">STORM-126</a>]</i>
